man/gpdasm.1
.TH GPDASM 1 "2007-10-27" "gputils-2.0.0_beta2" "Linux user's manual"
.SH NAME
gpdasm \- GNU PIC disassembler
.SH SYNOPSIS
.B gpdasm [options] file
.SH WARNING
The information in this man page is an extract from the full documentation of
gputils and is limited to the meaning of the options.  For complete and
current documentation, refer to "gputils.ps" located in the gputils docs
directory.
.SH DESCRIPTION
.B gpdasm
is an disassembler for Microchip (TM) PIC (TM) micro-controllers.
.B gpdasm
is part of gputils.  Check the
.BR gputils (1)
manpage for details on other GNU PIC utilities.
.SH OPTIONS
.TP
.BR \-c ", " \-\-mnemonics
Decode special mnemonics.
.TP
.BR \-h ", "\-\-help
Show the usage message.
.TP
.BR \-i ", "\-\-hex-info
Report extra information about input hex file.  This information includes file
size and hex format.
.TP
.BR \-l ", "\-\-list-chips
List the supported processors.
.TP
.BR \-m ", "\-\-dump
Display the contents of instruction memory.
.TP
.BR "\-p PROC" , " \-\-processor PROC
Select the processor.
.TP
.BR \-s ", "\-\-short
Output the short format.  This format can be reassembled by gpasm.
.TP
.BR \-v ", "\-\-version
Show the version.
.TP
.BR \-y ", "\-\-extended
Enable 18xx extended mode.
.SH SEE ALSO
.BR gputils (1)
.SH AUTHOR
Craig Franklin <craigfranklin@users.sourceforge.net>
.SH COPYRIGHT
Copyright (C) 2002, 2003, 2004, 2005 Craig Franklin

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


history.txt
======
snutil
======
Rev:		Date:		Reason:
0.0.3		12/6/29		eval version release	
0.1.0		12/6/29		sn16 class add 4 insn (inca, jza, deca, djza)
0.2.0		12/7/3		add cod output
0.3.0		12/7/6		add prefix S2-CONFIG BIT; S3-CHIP_ID in S19 output
0.4.0		12/7/10		sn16 class add 2 insn (rra, rla)
0.5.0		12/7/13		add coff output, add env param PATH 
0.6.0		12/8/22		correct "dt" assemble
0.6.1		12/10/23	add processor MC33P94, MC33P716, MC30P011
0.7.0		12/11/19	modify memonic rsubra, rsbcra etc.
0.8.0		12/12/5		sn16 class add 2 insn (lcall, lgoto); add processor MC32P21
0.8.1		12/12/14	correct lcall/lgoto opcode
0.9.0		13/1/20		correct lcall/lgoto link
1.0.0		13/3/19		add sn13 class; comra -> comr (sn14)
1.1.0		13/3/20		adcar -> adcra (sn13)
1.2.0		13/4/12		remove website
1.3.0		13/6/19		nessary update for c compiler
1.3.1		13/7/9		add processor MC32P64
1.4.0		13/9/16		correct relocate error; expand mc30 OPF7 insn address to 7bit
1.4.1_beta	13/10/15	add sn13b class, MC34P01
1.4.2_beta	14/1/6		sn13b opcode update, MC34P01
1.5.0		14/1/13		release sn13b class, MC34P01; remove adcai of sn16 class
1.6.0		14/1/22		add sn16b modify lcall/lgoto; remove lcall/lgoto of of other class; add processor MC32P821, MC33P78, MC33P116, MC30P081
1.6.1		14/1/23		correct gplink for lcall/lgoto
1.7.0_beta	14/4/11		add sn13c for MC31P5130
1.9.1
                加入新型号7323  之前MC32P7323用的是3221 在超空间1K报错。
        MC32P7323是8K的，超过8K报错，使用的是PRO_CLASS_16A内核，型号7323
1.9.0
                加入新内核 PRO_CLASS_SN16E      for     MC32T7333、MC9902
        MC32P7323是8K的，超过8K报错，使用的是PRO_CLASS_16A内核，型号7323
1.9.2	
		add processor 4K7323 for space 4K
	MC32P7323
1.9.3
		add processor 7212 for MC32P7212
1.9.4 	
		add processor 1K6220,05K6220 for MC30P6220	17/04/17
1.9.5
		add processor 1K7041,2K7041 for MC35P7041	17/04/18
		modify 7333 setup for MC32F7333			17/04/21
1.9.6
		add processor 7122	for MC32F7122		17/04/27
1.9.7		
		add processor 9903 9904 for A380		17/05/15	
1.9.8		
		add processor 7030 for MC32P7030		17/05/25
2.0.0_beta1 17/6/19	modify cod file for macro
2.0.0_beta2 17/6/23	output section location when link error ocurrs


gputils/gpdasm.c
/* Disassembles ".HEX" files
   Copyright (C) 2001, 2002, 2003, 2004, 2005
   Craig Franklin

This file is part of gputils.

gputils is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

gputils is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with gputils; see the file COPYING.  If not, write to
the Free Software Foundation, 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.  */

#include "stdhdr.h"

#include "libgputils.h"
#include "gpdasm.h"

char *processor_name = NULL;

struct gpdasm_state state = {
    no_processor,       /* processor type */
    PROC_CLASS_GENERIC, /* 12 bit device */
    1                   /* output format */
};

void select_processor(void)
{
  const struct px *found = NULL;

  if (processor_name == NULL)
  {
    printf("error: must select processor\n");
    exit(1);
  }

  found = gp_find_processor(processor_name);

  if (found)
  {
    state.processor = found;
  }
  else
  {
    printf("Didn't find any processor named: %s\nHere are the supported processors:\n",
           processor_name);
    gp_dump_processor_list(true, 0);
    exit(1);
  }

  state.class = gp_processor_class(state.processor);

  if (state.class->instructions == NULL)
  {
    fprintf(stderr, "error: unsupported processor class\n");
    exit(1);
  }

  return;
}

void writeheader()
{
  if (!state.format)
  {
    printf("\n");
    printf("        processor %s\n", processor_name);
  }
}

void closeasm()
{
  if (!state.format)
  {
    printf("        end\n");
  }
}

void writeorg(int org)
{
  if (!state.format)
  {
    printf("\n");
    printf("        org\t%#x\n", org);
  }
}

void dasm(MemBlock *memory)
{
  MemBlock *m = memory;
  int i, maximum;
  int last_loc = 0;
  int num_words;
  char buffer[80];

  writeheader();

  while (m)
  {
    i = m->base << I_MEM_BITS;

    maximum = i + MAX_I_MEM;

    while (i < maximum)
    {
      unsigned short data;
      if (state.class->i_memory_get(memory, i, &data))
      {
        int org = gp_processor_byte_to_org(state.class, i);
        if (last_loc != i - 2)
        {
          writeorg(org);
        }
        last_loc = i;
        if (state.format)
        {
          printf("%06x:  %04x  ", org, data);
        }
        else
        {
          printf("        ");
        }
        num_words = gp_disassemble(memory,
                                   i,
                                   state.class,
                                   buffer,
                                   sizeof(buffer));
        printf("%s\n", buffer);
        if (num_words != 1)
        {
          i += 2;
          /* some 18xx instructions use two words */
          if (state.format)
          {
            state.class->i_memory_get(memory, i, &data);
            printf("%06x:  %04x\n", gp_processor_byte_to_org(state.class, i), data);
          }
        }
      }
      i += 2;
    }

    m = m->next;
  }

  closeasm();
}

void show_usage(void)
{
  printf("Usage: gpdasm [options] file\n");
  printf("Options: [defaults in brackets after descriptions]\n");
  printf("  -c, --mnemonics                Decode special mnemonics.\n");
  printf("  -h, --help                     Show this usage message.\n");
  printf("  -i, --hex-info                 Information on input hex file.\n");
  printf("  -l, --list-chips               List supported processors.\n");
  printf("  -m, --dump                     Memory dump hex file.\n");
  printf("  -p PROC, --processor PROC      Select processor.\n");
  printf("  -s, --short                    Print short format.\n");
  printf("  -v, --version                  Show version.\n");
  /*  printf("  -y, --extended                 Enable 18xx extended mode.\n");
  printf("      --strict                   Disassemble only opcodes generated by gpasm"
         "                                 in case of instructions with serveral opcodes\n");*/
  printf("\n");
  //printf("Report bugs to:\n");
  //printf("%s\n", PACKAGE_BUGREPORT);
  exit(0);
}

#define GET_OPTIONS "?chilmp:svy"

/* Used: himpsv */
static struct option longopts[] =
    {
        {"mnemonics", 0, 0, 'c'},
        {"help", 0, 0, 'h'},
        {"hex-info", 0, 0, 'i'},
        {"list-chips", 0, 0, 'l'},
        {"dump", 0, 0, 'm'},
        {"processor", 1, 0, 'p'},
        {"short", 0, 0, 's'},
        {"version", 0, 0, 'v'},
        {"extended", 0, 0, 'y'},
        {"strict", 0, 0, 't'},
        {0, 0, 0, 0}};

#define GETOPT_FUNC getopt_long(argc, argv, GET_OPTIONS, longopts, 0)

int main(int argc, char *argv[])
{
  extern char *optarg;
  extern int optind;
  int c;
  int print_hex_info = 0;
  int usage = 0;
  int memory_dump = 0;
  char *filename = 0;
  int strict = false;
  gp_init();

  state.i_memory = i_memory_create();

  while ((c = GETOPT_FUNC) != EOF)
  {
    switch (c)
    {
    case '?':
    case 'h':
      usage = 1;
      break;
    case 'c':
      gp_decode_mnemonics = true;
      break;
    case 'i':
      print_hex_info = 1;
      break;
    case 'l':
      gp_dump_processor_list(true, 0);
      exit(0);
      break;
    case 'm':
      memory_dump = 1;
      break;
    case 'p':
      processor_name = optarg;
      break;
    case 's':
      state.format = 0;
      break;
    /*    case 'y':
      gp_decode_extended = true;
      break;*/
    case 'v':
      fprintf(stderr, "%s\n", GPDASM_VERSION_STRING);
      exit(0);
    case 't':
      strict = true;
    }
    if (usage)
      break;
  }

  if ((optind + 1) == argc)
  {
    filename = argv[optind];
  }
  else
  {
    usage = 1;
  }

  if (usage)
  {
    show_usage();
  }

  select_processor();

  state.hex_info = readhex_s19(filename, state.i_memory);

  if (state.hex_info->error)
  {
    state.num.errors++;
  }

  if (strict && state.class && state.class->patch_strict)
  {
    state.class->patch_strict();
  }

  if (print_hex_info)
  {
    printf("s19 file name:   %s\n", filename);
    /*    printf("hex file format: ");
    if (state.hex_info->hex_format == inhx8m) {
      printf("inhx8m\n");
    } else if (state.hex_info->hex_format == inhx16) {
      printf("inhx16\n");
    } else if (state.hex_info->hex_format == inhx32) {
      printf("inhx32\n");
    } else {
      printf("UNKNOWN\n");
    }*/
    printf("number of bytes: %i\n", state.hex_info->size);
    printf("\n");
  }

  if (state.num.errors == 0)
  {
    if (memory_dump)
    {
      print_i_memory(state.i_memory, state.class);
    }
    else
    {
      dasm(state.i_memory);
    }
  }

  i_memory_free(state.i_memory);

  if (state.num.errors > 0)
    return EXIT_FAILURE;
  else
    return EXIT_SUCCESS;
}
